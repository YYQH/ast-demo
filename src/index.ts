import * as t from '@babel/types';

// export interface UsePoolsQuery<TData> extends ReactQueryParams<QueryPoolsResponse, TData> {
//   request?: QueryPoolsRequest;
// }
const createInterfaceDeclaration = ({
  queryInterface = 'UsePoolsQuery',
  requestType = 'QueryPoolsRequest',
  responseType = 'QueryPoolsResponse',
}) => {
  // interface UsePoolsQuery
  const interfaceName = t.identifier(queryInterface);
  // <TData>
  const typeParameters = t.tsTypeParameterDeclaration([
    t.tsTypeParameter(null, null, 'TData')
  ]);
  // <QueryPoolsResponse, TData>
  const reactQueryParamsType = t.tsTypeParameterInstantiation([
    t.tsTypeReference(
      t.identifier(responseType)
    ),
    t.tsTypeReference(
      t.identifier('TData')
    )
  ])
  // extends ReactQueryParams
  const interfaceExtends = t.tsExpressionWithTypeArguments(
    t.identifier('ReactQueryParams'),
    reactQueryParamsType
  )
  // request?: QueryPoolsRequest;
  const interfaceBody = t.tsInterfaceBody([
    t.tsPropertySignature(
      t.identifier('request'),
      t.tsTypeAnnotation(
        t.tsTypeReference(
          t.identifier(requestType)
        )
      )
    ),

  ])
  return t.exportNamedDeclaration(
    t.tsInterfaceDeclaration(
      interfaceName,
      typeParameters,
      [interfaceExtends],
      interfaceBody
    )
  )
}

// const usePools = <TData = QueryPoolsResponse,>({
//   request,
//   options
// }: UsePoolsQuery<TData>) => {
//   return useQuery<QueryPoolsResponse, Error, TData>(["poolsQuery", request], () => {
//       if (!queryService) throw new Error("Query Service not initialized");
//       return queryService.pools(request);
//   }, options);
// };
const createFunctionDeclaration = ({
  queryInterface = 'UsePoolsQuery',
  hookName = 'usePools',
  requestType = 'QueryPoolsRequest',
  responseType = 'QueryPoolsResponse',
  queryServiceMethodName = 'pools',
  queryKey = 'poolsQuery'
}) => {
  // ({ request, options })
  const arrowFunctionParams = t.objectPattern(
    [
      t.objectProperty(
        t.identifier('request'),
        t.identifier('request')
      ),
      t.objectProperty(
        t.identifier('options'),
        t.identifier('options'),
      )
    ]
  )
  // ({ ... }): UsePoolsQuery<TData>
  arrowFunctionParams.typeAnnotation = t.tsTypeAnnotation(
    t.tsTypeReference(
      t.identifier(queryInterface),
      t.tsTypeParameterInstantiation([
        t.tsTypeReference(
          t.identifier('TData')
        )
      ])
    )
  )
  const returnCallArrowFunction = t.callExpression(
    // useQuery
    t.identifier('useQuery'),
    [
      // ["poolsQuery", request]
      t.arrayExpression([
        t.stringLiteral(queryKey),
        t.identifier('request')
      ]),
      // () => {}
      t.arrowFunctionExpression(
        [],
        t.blockStatement([
          // if (!queryService) throw new Error("Query Service not initialized");
          t.ifStatement(
            t.unaryExpression(
              '!',
              t.identifier('queryService')
            ),
            t.throwStatement(
              t.newExpression(
                t.identifier('Error'),
                [
                  t.stringLiteral('Query Service not initialized')
                ]
              )
            )
          ),
          // return queryService.pools(request);
          t.returnStatement(
            t.callExpression(
              t.memberExpression(
                t.identifier('queryService'),
                t.identifier(queryServiceMethodName)
              ),
              [
                t.identifier('request')
              ]
            )
          )
        ])
      ),
      // options
      t.identifier('options')
    ],
  )
  // <QueryPoolsResponse, Error, TData>
  returnCallArrowFunction.typeParameters = t.tsTypeParameterInstantiation(
    [
      t.tsTypeReference(
        t.identifier(responseType)
      ),
      t.tsTypeReference(
        t.identifier('Error')
      ),
      t.tsTypeReference(
        t.identifier('TData')
      )
    ]
  )
  // arrow function body
  const arrowFunctionBody = t.blockStatement(
    // arrow function body
    [
      // return
      t.returnStatement(
        returnCallArrowFunction,
      )
    ]
  )
  // arrow function
  const arrayFunction = t.arrowFunctionExpression(
    [
      arrowFunctionParams
    ],
    arrowFunctionBody
  )
  // <TData = QueryPoolsResponse,>
  arrayFunction.typeParameters = t.tsTypeParameterDeclaration([
    t.tsTypeParameter(
      null,
      t.tsTypeReference(
        t.identifier(responseType)
      ),
      'TData'
    )
  ])
  // const usePools 
  const variableName = t.variableDeclarator(
    t.identifier(hookName),
    arrayFunction
  )
  return t.variableDeclaration('const', [variableName])
}

export function generateHookCode(config) {


  return t.file(
    t.program([
      createInterfaceDeclaration(config),
      createFunctionDeclaration(config)
    ])
  )
}