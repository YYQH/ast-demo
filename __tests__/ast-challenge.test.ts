import { generateHookCode } from '../src';
import generate from '@babel/generator';

describe('AST challenge test', () => {
  it('should generate the hook correctly', () => {
    const hookConfig = {
      queryInterface: 'UsePoolsQuery',
      hookName: 'usePools',
      requestType: 'QueryPoolsRequest',
      responseType: 'QueryPoolsResponse',
      queryServiceMethodName: 'pools',
      queryKey: 'poolsQuery'
    }
    const ast = generateHookCode(hookConfig);
    const { code } = generate(ast);
    console.log(111, code);
  })

  const showResult = json => {
    Object.entries(json).forEach(([methodName, types ]) => {
      const hookConfig = {
        queryInterface: `Use${methodName}Query`,
        hookName: `use${methodName}`,
        requestType: types.requestType,
        responseType: types.responseType,
        queryServiceMethodName: (methodName as string).toLowerCase(),
        queryKey: `${(methodName as string).toLowerCase()}Query`
      }
      const ast = generateHookCode(hookConfig);
      const { code } = generate(ast);
      console.log(111, code);
      
      expect(code).toMatchSnapshot();
    })
  }
  
  it('Take the input and output below, and make a function that can generate this code using babel AST:', () => {
    const json = {
      "Pools": {
          "requestType": "QueryPoolsRequest",
          "responseType": "QueryPoolsResponse"
      }
    };
    showResult(json)
  })
  it('test example-methods.json', () => {
    const exampleJson = require('../example-methods.json');
    // console.log(exampleData);
    showResult(exampleJson)
  })
})