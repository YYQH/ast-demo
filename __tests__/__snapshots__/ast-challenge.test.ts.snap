// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AST challenge test Take the input and output below, and make a function that can generate this code using babel AST: 1`] = `
"export interface UsePoolsQuery<TData> extends ReactQueryParams<QueryPoolsResponse, TData> {
  request: QueryPoolsRequest;
}

const usePools = <TData = QueryPoolsResponse,>({
  request: request,
  options: options
}: UsePoolsQuery<TData>) => {
  return useQuery<QueryPoolsResponse, Error, TData>([\\"poolsQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.pools(request);
  }, options);
};"
`;

exports[`AST challenge test should generate the hook correctly 1`] = `"interface UsePoolsQuery {}"`;

exports[`AST challenge test test example-methods.json 1`] = `
"export interface UsePoolsQuery<TData> extends ReactQueryParams<QueryPoolsResponse, TData> {
  request: QueryPoolsRequest;
}

const usePools = <TData = QueryPoolsResponse,>({
  request: request,
  options: options
}: UsePoolsQuery<TData>) => {
  return useQuery<QueryPoolsResponse, Error, TData>([\\"poolsQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.pools(request);
  }, options);
};"
`;

exports[`AST challenge test test example-methods.json 2`] = `
"export interface UseNumPoolsQuery<TData> extends ReactQueryParams<QueryNumPoolsResponse, TData> {
  request: QueryNumPoolsRequest;
}

const useNumPools = <TData = QueryNumPoolsResponse,>({
  request: request,
  options: options
}: UseNumPoolsQuery<TData>) => {
  return useQuery<QueryNumPoolsResponse, Error, TData>([\\"numpoolsQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.numpools(request);
  }, options);
};"
`;

exports[`AST challenge test test example-methods.json 3`] = `
"export interface UseTotalLiquidityQuery<TData> extends ReactQueryParams<QueryTotalLiquidityResponse, TData> {
  request: QueryTotalLiquidityRequest;
}

const useTotalLiquidity = <TData = QueryTotalLiquidityResponse,>({
  request: request,
  options: options
}: UseTotalLiquidityQuery<TData>) => {
  return useQuery<QueryTotalLiquidityResponse, Error, TData>([\\"totalliquidityQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.totalliquidity(request);
  }, options);
};"
`;

exports[`AST challenge test test example-methods.json 4`] = `
"export interface UsePoolQuery<TData> extends ReactQueryParams<QueryPoolResponse, TData> {
  request: QueryPoolRequest;
}

const usePool = <TData = QueryPoolResponse,>({
  request: request,
  options: options
}: UsePoolQuery<TData>) => {
  return useQuery<QueryPoolResponse, Error, TData>([\\"poolQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.pool(request);
  }, options);
};"
`;

exports[`AST challenge test test example-methods.json 5`] = `
"export interface UsePoolParamsQuery<TData> extends ReactQueryParams<QueryPoolParamsResponse, TData> {
  request: QueryPoolParamsRequest;
}

const usePoolParams = <TData = QueryPoolParamsResponse,>({
  request: request,
  options: options
}: UsePoolParamsQuery<TData>) => {
  return useQuery<QueryPoolParamsResponse, Error, TData>([\\"poolparamsQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.poolparams(request);
  }, options);
};"
`;

exports[`AST challenge test test example-methods.json 6`] = `
"export interface UseTotalPoolLiquidityQuery<TData> extends ReactQueryParams<QueryTotalPoolLiquidityResponse, TData> {
  request: QueryTotalPoolLiquidityRequest;
}

const useTotalPoolLiquidity = <TData = QueryTotalPoolLiquidityResponse,>({
  request: request,
  options: options
}: UseTotalPoolLiquidityQuery<TData>) => {
  return useQuery<QueryTotalPoolLiquidityResponse, Error, TData>([\\"totalpoolliquidityQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.totalpoolliquidity(request);
  }, options);
};"
`;

exports[`AST challenge test test example-methods.json 7`] = `
"export interface UseTotalSharesQuery<TData> extends ReactQueryParams<QueryTotalSharesResponse, TData> {
  request: QueryTotalSharesRequest;
}

const useTotalShares = <TData = QueryTotalSharesResponse,>({
  request: request,
  options: options
}: UseTotalSharesQuery<TData>) => {
  return useQuery<QueryTotalSharesResponse, Error, TData>([\\"totalsharesQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.totalshares(request);
  }, options);
};"
`;

exports[`AST challenge test test example-methods.json 8`] = `
"export interface UseSpotPriceQuery<TData> extends ReactQueryParams<QuerySpotPriceResponse, TData> {
  request: QuerySpotPriceRequest;
}

const useSpotPrice = <TData = QuerySpotPriceResponse,>({
  request: request,
  options: options
}: UseSpotPriceQuery<TData>) => {
  return useQuery<QuerySpotPriceResponse, Error, TData>([\\"spotpriceQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.spotprice(request);
  }, options);
};"
`;

exports[`AST challenge test test example-methods.json 9`] = `
"export interface UseEstimateSwapExactAmountInQuery<TData> extends ReactQueryParams<QuerySwapExactAmountInResponse, TData> {
  request: QuerySwapExactAmountInRequest;
}

const useEstimateSwapExactAmountIn = <TData = QuerySwapExactAmountInResponse,>({
  request: request,
  options: options
}: UseEstimateSwapExactAmountInQuery<TData>) => {
  return useQuery<QuerySwapExactAmountInResponse, Error, TData>([\\"estimateswapexactamountinQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.estimateswapexactamountin(request);
  }, options);
};"
`;

exports[`AST challenge test test example-methods.json 10`] = `
"export interface UseEstimateSwapExactAmountOutQuery<TData> extends ReactQueryParams<QuerySwapExactAmountOutResponse, TData> {
  request: QuerySwapExactAmountOutRequest;
}

const useEstimateSwapExactAmountOut = <TData = QuerySwapExactAmountOutResponse,>({
  request: request,
  options: options
}: UseEstimateSwapExactAmountOutQuery<TData>) => {
  return useQuery<QuerySwapExactAmountOutResponse, Error, TData>([\\"estimateswapexactamountoutQuery\\", request], () => {
    if (!queryService) throw new Error(\\"Query Service not initialized\\");
    return queryService.estimateswapexactamountout(request);
  }, options);
};"
`;
